### PR TITLE
NF:  q-space trajectory imaging

### PR DESCRIPTION
**Pull request by kerkelae**
Hi everyone,

I am working on an implementation of q-space trajectory imaging (https://doi.org/10.1016/j.neuroimage.2016.02.039). Since I have not implemented anything in the `dipy.reconst` module before, I created this pull request way before the code is finished to confirm that my approach follows Dipy conventions and to avoid having to rewrite too much code.

So far, I have implemented first versions of the classes `QtiModel` and `QtiFit` that include an OLS fit and a few parameter maps. The code snippet below shows an example of the user interface:
```python
# Specify b-tensor shapes when generating the gradient table
btens = np.array(['LTE' for i in range(107)] + ['STE' for i in range(107)])
gtab = gradient_table(bvals, bvecs, btens=btens)

# Fit QTI
qtimodel = qti.QtiModel(gtab, 'OLS')
qtifit = qtimodel.fit(data, mask)

# Calculate microscopic fractional anisotropy map
uFA = qtifit.uFA
```

In the next few weeks, I aim to do the following:
- Efficient WLS fit (and tests)
- Methods for calculating the parameter maps (and tests)
- Share my MDE data to be used as an example dataset and write a data fetcher for it by following the examples in `dipy.data.fetcher`
- Tutorial

All feedback and suggestions are very much appreciated!

### CODE REVIEW

**Comment by skoudoro** [comment]
Great! Thank you for doing this @kerkelae!

Do you want us to review your code at each step or we just keep an eye until you tell us that you are ready for a review?

**Comment by kerkelae** [comment]
@skoudoro I'll give you an update here when I feel like the code is ready for review and I'd like to get some comments on how to improve it.

File: dipy/reconst/qti.py [hunk 1]
```diff
@@ -0,0 +1,652 @@
...
+from warnings import warn
+
+import numpy as np
+
+from dipy.reconst.base import ReconstModel
+from dipy.core.ndindex import ndindex
+from dipy.reconst.dti import auto_attr
+
+
+def from_3x3_to_6x1(T):
```
Comment (arokem) [inline dipy/reconst/qti.py:13]: Looks a lot like: https://github.com/dipy/dipy/blob/master/dipy/reconst/dti.py#L1956

File: dipy/reconst/qti.py [hunk 2]
```diff
@@ -0,0 +1,652 @@
...
+         [C * V[..., 8, 0], C * V[..., 11, 0], C * V[..., 14, 0],
+          C * V[..., 20, 0], C * V[..., 19, 0], V[..., 17, 0]]))
+    T = np.moveaxis(T, (0, 1), (-2, -1))
+    return T
+
+
+# These tensors are used in the calculation of QTI parameters
+e_iso = np.eye(3) / 3
+E_iso = np.eye(6) / 3
+E_bulk = np.matmul(from_3x3_to_6x1(e_iso), from_3x3_to_6x1(e_iso).T)
```
Comment (arokem) [inline dipy/reconst/qti.py:128]: Maybe this would improve readability?  

```suggestion
E_bulk = from_3x3_to_6x1(e_iso) @ from_3x3_to_6x1(e_iso).T
```

Comment (arokem) [reply dipy/reconst/qti.py:128]: There are a few other places where `np.matmul` is used, where I think readability would improve by using `@`

File: dipy/reconst/qti.py [hunk 3]
```diff
@@ -0,0 +1,652 @@
...
+    -------
+    params : numpy.ndarray
+        Array of shape (..., 28) containing the estimated model parameters.
+        Element 0 is the estimated signal without diffusion-weighting, elements
+        1-6 are the estimated diffusion tensor parameters, and elements 7-27 are
+        the estimated covariance tensor parameters.
+    """
+    params = np.zeros(mask.shape + (28,)) * np.nan
+    index = ndindex(mask.shape)
+    for v in index:  # This loop is slow loop
```
Comment (arokem) [inline dipy/reconst/qti.py:316]: Check out https://github.com/dipy/dipy/blob/master/dipy/reconst/dti.py#L1224 for our approach there. 

See also: https://github.com/dipy/dipy/blob/master/dipy/reconst/dti.py#L1442 for the solve itself.

And check out https://github.com/dipy/dipy/blob/c846bf5a23b7e95343a9cf231df2653473602456/dipy/utils/arrfuncs.py#L29 for vectorized pinv

**Review by arokem — COMMENTED** [review]
Just a couple of comments for now. 

One general comment: Would be great to add `predict` methods (to both classes) to use in calculating goodness of fit and so on.

File: dipy/reconst/qti.py (hunk 3 above)
Comment (ShreyasFadnavis) [reply dipy/reconst/qti.py:316]: @kerkelae : might also be worthwhile to compare the above `pinv` with the SciPy `pinv2` both seem to be doing something similar. Also, the SciPy SVD is in general faster than NumPy. Also, it allows you to set `check_finite=False`. 

https://docs.scipy.org/doc/scipy/reference/generated/scipy.linalg.pinv2.html

Maybe we should change the SVD in `arrfuncs.py` too?

**Comment by kerkelae** [comment]
I have implemented the first versions of the features we discussed in the Dipy lab meeting two weeks ago: tutorial in `doc/examples/reconst_qti.py`, data fetcher in `dipy/data/fetcher.py`, and parameter maps and fit optimization in `dipy/reconst/qti.py`. I would like to ask you to have a look at the code in its current state because I will have to focus on other things during the next few weeks. I will have time to finish this project after May. I feel the the fundamentals are there, but the some parts of the code need cleaning up and more unit tests are needed. @RafaelNH mentioned that he would be interested in implementing some more complicated unit tests to validate the code.

**Comment by Garyfallidis** [comment]
@RafaelNH and all can you look at this PR and provide feedback to @kerkelae for next steps?

**Comment by kerkelae** [comment]
Any updates on this from @RafaelNH or @DenebBoito?

**Review by RafaelNH — CHANGES_REQUESTED** [review]
Hi @kerkelae! Amazing work! I think this is almost ready to be merged. 

In addition to my code review, I was also able to run your tutorial. I've noticed that when fitting the tutorials data, the warning message "All matrices converted to Voigt notation are not symmetric". What part of your data is inducing the massage - background voxels? Not sure if this message is providing enough information for users to understand want is going on. Is this message really necessary given that it is raised even on the tutorial's idea dataset?

Please, check that your tests cover all lines of code in qti.py. Lines of codes that you are not tested: 274, 359-362, 402-408, 473-475, 478, 539, 901-906. 

Lastly, do not forget to rebase this PR.

**Comment by kerkelae** [comment]
@skoudoro I'll be able to join in for 15-20 min in the beginning of the meeting if we discuss this PR first. Would that work?

**Comment by Garyfallidis** [comment]
@kerkelae The docstrings in the example have no line spacings and therefore will not render correctly. Have you tried to compile the documentation? Make sure you rebase first. Apart from those small things this is ready to be merged.

**Comment by Garyfallidis** [comment]
To render the specific example go to 
``
cd dipy/doc/examples
``
and disable all unrelated examples in ``valid_examples.txt`` by replacing space `` ``  with ``# `` in front.

``
cd ..

make clean

make html
``
You will also need sphinx ``pip install sphinx``.

**Comment by kerkelae** [comment]
@Garyfallidis I added the empty lines and built the documentation locally. I think it works now.

**Comment by kerkelae** [comment]
Thanks @skoudoro . I made the changes you requested and ran the tests.

**Comment by skoudoro** [comment]
Thank you @kerkelae. I think this PR is ready to go. I will just check the documentation generation after that I will go ahead and merge it.

Is there any additional comment @Garyfallidis? @arokem, @ShreyasFadnavis, or @RafaelNH?

**Comment by Garyfallidis** [comment]
Congratulations @kerkelae !!! Great work!  :1st_place_medal: Can't wait to hear about this work in the DIPY workshop! https://dipy.org/workshops/dipy-workshop-2022

**Pull request merged**